Migrate shopper free trial subscription purchase to Playwright

tpaksu opened the pull request:
Fixes #10086


#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

* This PR adds a new shopper subscriptions purchase free trial spec file to the Playwright test directory.
* This PR also removes the old shopper subscription free trial purchase test file from the Puppeteer test directory.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Checkout this branch
* Run `./tests/e2e-pw/test-e2e-pw.sh shopper-subscriptions-purchase-free`
* All tests should pass.
* All tests on GH should pass.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

tpaksu committed:
Migrate shopper free trial subscription purchase to Playwright

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10159` or branch name `dev/10086-playwright-migration-shopper-subscriptions-purchase-free-trial` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/10086-playwright-migration-shopper-subscriptions-purchase-free-trial) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c6ac2c66ed7440f19412d8666a8c1e7a298dd759
- Build time: 2025-01-22 10:08:51 UTC

_Note: the build is updated when a new commit is pushed to this PR._

tpaksu committed:
Merge branch 'develop' into dev/10086-playwright-migration-shopper-subscriptions-purchase-free-trial

tpaksu committed:
Remove duplicates

tpaksu committed:
Merge branch 'develop' into dev/10086-playwright-migration-shopper-subscriptions-purchase-free-trial

eduardoumpierre commented on tests/e2e-pw/specs/shopper/shopper-subscriptions-purchase-free-trial.spec.ts:
```diff
@@ -0,0 +1,171 @@
+		test( 'Merchant should be able to purchase a free trial', async ( {
+			browser,
+		} ) => {
+			const { shopperPage } = await getShopper( browser );
```
Since you're deleting `subscriptions-customer`, it’s better to use `getAnonymousShopper()` instead, as `getShopper()` logs in with a different user (the regular customer).

tpaksu replied:
Actually, this test (and IMO most of the tests) doesn't need deleting the customer, we can go on with the regular customer. I just copied what was there in the puppeteer test. What about we deleting the regular customer instead to reset the saved cards? Did you understand the reason why we're deleting the customers?

tpaksu replied:
@eduardoumpierre any follow-ups to my above statement?

tpaksu replied:
Ok, just realized that the test doesn't need customer deletion, changed the process to use the standard customer. You were right about the first comment by the way, it didn't have any effect on the regular shopper. Thanks for bringing that up!

allie500 commented on tests/e2e-pw/specs/shopper/shopper-subscriptions-purchase-free-trial.spec.ts:
```diff
@@ -0,0 +1,171 @@
+const customerBilling = config.addresses.customer.billing;
+const customerBillingConfig =
+	config.addresses[ 'subscriptions-customer' ].billing;
+let orderId, subscriptionId;
```
Nitpick: let's set these variables' types.

```
let orderId: string, subscriptionId: string;
```

tpaksu replied:
Thanks for the reminder, I've set them in the last iteration.

allie500 commented on tests/e2e-pw/specs/shopper/shopper-subscriptions-purchase-free-trial.spec.ts:
```diff
@@ -0,0 +1,171 @@
+			).toHaveText( /\(seti_.*\)/ );
+
+			await goToSubscriptions( merchantPage );
+			const subscriptionRow = await merchantPage.locator(
```
Nitpick: we can remove `await`, it isn't needed here.

```
'await' has no effect on the type of this expression.ts(80007)
```

tpaksu replied:
Thanks, fixed it.

allie500 reviewed (commented):
Thanks for working on this spec, Taha. The code reads well and the tests are passing for me. I left a couple nitpicks, but other than that it looks good to me.

Do we need to activate or make sure that multi-currency wasn't activated like we've been doing in some other tests?

I'll wait to approve based on your answer to the above question.

tpaksu committed:
Merge branch 'develop' into dev/10086-playwright-migration-shopper-subscriptions-purchase-free-trial

tpaksu committed:
Simplify test by removing customer deletion and class locators

tpaksu commented:
@allie500 I've changed most of code to remove the class based locators and use user-centric selectors, also removed the customer deletion part, which this test didn't need at all. If you can re-review from scratch, I'll appreciate!

tpaksu committed:
Merge branch 'develop' into dev/10086-playwright-migration-shopper-subscriptions-purchase-free-trial

eduardoumpierre reviewed (approved):
The test spec is passing and the code looks good. Thanks for working on this one!

allie500 reviewed (approved):
Thanks for making the updates. The code looks great and the tests are passing. Let's ship it!

tpaksu commented:
![](https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExcTFrcTdvd2UwNTRzM2g2c2Ewa3dnb29icXZidjR5YTM5eTZkMnYwciZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/GNtj4JoNwDkas/giphy.gif)
